### PR TITLE
Add Mov64 Support

### PR DESCRIPTION
**Pull request by Negroni1**

### CODE REVIEW

**Comment by thatjiaozi** [comment]
Plz run the format script and upload a commit with those changes :)

File: pkg/ebpf/alu_instructions_test.go [hunk 1]
```diff
@@ -737,3 +737,68 @@ func TestAluInstructionGenerationAndEncoding(t *testing.T) {
...
+		if opcode.Mode != pb.StLdMode_StLdModeIMM {
+			t.Fatalf("operationCode = %d, want %d", opcode.Mode, pb.StLdMode_StLdModeIMM)
+		}
+
+		encodingArray, err := encodeInstruction(instruction)
+		if err != nil {
+			t.Fatalf("unexpected error when ecoding: %v", err)
+		}
+
+		if !reflect.DeepEqual(encodingArray, []uint64{0x9abcdef000000018, 0x1234567800000000}) {
```
Comment (thatjiaozi) [inline pkg/ebpf/alu_instructions_test.go:800]: Maybe mention that this encoding was manually computed

**Review by thatjiaozi — CHANGES_REQUESTED** [review]
First batch of comments

File: pkg/strategies/playground.go [hunk 2]
```diff
@@ -21,7 +21,7 @@ type Playground struct {
 // GenerateProgram should return the instructions to feed the verifier.
 func (pg *Playground) GenerateProgram(ffi *units.FFI) (*epb.Program, error) {
 	insn, err := InstructionSequence(
-		Mov64(R0, 0),
+		Mov64(R0, int64(0x123456789ABCDEF0)),
```
[the remark this answers is not in the document]
Comment (Negroni1) [reply pkg/strategies/playground.go]: Changed it such that it verifies if its a int64, but directly to the Mov64 instruction. Lmk what you think

```
if int(int32(src)) != int(int64(src)) {                                     
    return newAluInstruction(pb.AluOperationCode_AluMov, pb.InsClass_InsClassAlu64, dstReg, int64(src))
}   
```

File: pkg/ebpf/alu_instructions.go [hunk 3]
```diff
@@ -191,6 +230,12 @@ func Xor[T Src](dstReg pb.Reg, src T) *pb.Instruction {
 // Mov64 Creates a new 64 bit Mov instruction that is either imm or reg depending
 // on the data type of src
 func Mov64[T Src](dstReg pb.Reg, src T) *pb.Instruction {
+	// The ALU instruction requires explicit indication of int64 if the
+	// instruction has a 64-bit immediate value. Otherwise, it will interpret
+	// the value as an int
+	if int(int32(src)) != int(int64(src)) {
```
Comment (thatjiaozi) [inline pkg/ebpf/alu_instructions.go:236]: Could we add a test case where we make sure that int32 -> emits an alu mov, int64 -> emits a load instruction?

**Review by thatjiaozi — CHANGES_REQUESTED** [review]

File: pkg/ebpf/alu_instructions.go (hunk 3 above)
Comment (Negroni1) [reply pkg/ebpf/alu_instructions.go:236]: I think there is already a test case for it:
https://github.com/google/buzzer/blob/edab5305d4ca6af0dcbb3b935c3cd39151c84b6d/pkg/ebpf/alu_instructions_test.go#L308C4-L308C73

File: pkg/ebpf/alu_instructions.go [hunk 4]
```diff
@@ -32,6 +32,45 @@ func newAluInstruction[T Src](oc pb.AluOperationCode, insclass pb.InsClass, dst
 		srcReg = pb.Reg_R0
 		intImm := any(src).(int)
 		imm = int32(intImm)
+		/*
+		   Currenty, only the Mov64 instruction in the ALU instruction set handles
+		   64-bit immediate values. It achieves this by using the wide instruction encoding.
+
+		   https://docs.kernel.org/bpf/standardization/instruction-set.html#bit-immediate-instructions
+		*/
+	case int64:
```
[the remark this answers is not in the document]
Comment (Negroni1) [reply pkg/ebpf/alu_instructions.go:41]: https://github.com/google/buzzer/pull/61/commits/2567a946b5c1cbb5b7144588d77c0e3ebaeab61c

File: pkg/ebpf/alu_instructions.go [hunk 5]
```diff
@@ -39,37 +39,39 @@ func newAluInstruction[T Src](oc pb.AluOperationCode, insclass pb.InsClass, dst
...
-					DstReg:    0,
-					SrcReg:    0,
-					Offset:    0,
-					Immediate: upper,
-					PseudoInstruction: &pb.Instruction_Empty{
-						Empty: &pb.Empty{},
 					},
 				},
-			},
+			}
```
Comment (thatjiaozi) [inline pkg/ebpf/alu_instructions.go:74]: should we put an else clause here where you set the value of imm to the int32 cast of src?

**Review by thatjiaozi — APPROVED** [review]

File: pkg/ebpf/alu_instructions.go (hunk 5 above)
Comment (Negroni1) [reply pkg/ebpf/alu_instructions.go:74]: So if an instruction is generated (that is not Mov64) with a int64 immediate it gets treated as a int32 immediate?

**Review by amlweems — APPROVED** [review]

**Pull request merged**